portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -106,6 +114,23 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
     response.sendRedirect("/index.html");
   }
 
+  private boolean doesLimitExceeded(int limit) {
```

Comment:
Just FYI: this also will not stop hackers from spending your budget, since they can just translate everything many times. But it cases some problems indeed!

Reply:
Yes you are right :(
Do you have any advice about preventing multiple translation?

Reply:
So I see 3 options, I ordered them in the order of time required, choose any of them:

1) Leave it as is :) Every time the user translates everything, there is some delay 2-3 seconds which makes impossible quick DDOS attacks. But if someone decides to click "translate" for every 3 seconds during the rest of your internship, the cost may be (if there are a lot of comments):
20 (cost of million characters) * 10000 (total limit) * 60 (days until the end of the internship)  * 24 * 60 * 60 / (3 (click every 3 seconds) * 1000000) =345600 dollars. Sorry for calculating that :) While the total comment size is not big, it's fine to keep it as is.

2) Create the variable with the total number of translated characters. The question is how to store it :)
2.1) Use datastore
2.2) Try to just store the variable in the servlet class, initialise it at the class creation and decrease every time. I'm not sure that it will work though. Probably it will be the limit only for one user, one thread and until you restart the server. But it's better than nothing. Because I don't know when the DataServlet class gets created.
2.3) Use a ServletContext for storing the variable of the total number of characters that were ever translated: https://stackoverflow.com/questions/1832691/sharing-a-static-object-between-a-servlet-and-a-webservice
That's just a first link in google for "how to store the variable in servlet", I'm not sure it will work.

3) (preferred) Schedule an appointment with dev coaches and ask them! Or just sent an email.

Reply:
Thanks a lot :)
I think, for now it's not a big problem because it's now allowed only google accounts.